src/rmc.rs

```diff
@@ -0,0 +1,426 @@
+    FutureExt, StreamExt,
+};
+use log::debug;
+use std::{
+    cmp::Ordering,
+    collections::{hash_map::Entry, BinaryHeap, HashMap},
+    hash::Hash,
+};
+use tokio::time;
+
+/// A message consists of either a signed (indexed) hash, or a multisigned hash together with a
+/// signature of the multisignature by the sender.
+#[derive(Clone)]
+pub enum Message<H: Signable, S: Signature, M: PartialMultisignature> {
+    SignedHash(UncheckedSigned<Indexed<H>, S>),
+    MultisignedHash(UncheckedSigned<Indexed<UncheckedSigned<H, M>>, S>),
+}
+
+impl<H: Signable, S: Signature, M: PartialMultisignature> Message<H, S, M> {
+    fn hash(&self) -> &H {
+        match self {
+            Message::SignedHash(unchecked) => unchecked.as_signable().as_signable(),
+            Message::MultisignedHash(unchecked) => {
+                unchecked.as_signable().as_signable().as_signable()
+            }
+        }
+    }
+}
+
+pub struct IncomingMessage<H: Signable, MK: MultiKeychain> {
+    message: Message<H, MK::Signature, MK::PartialMultisignature>,
+    sender: NodeIndex,
+}
+
+pub struct OutgoingMessage<H: Signable, MK: MultiKeychain> {
+    pub message: Message<H, MK::Signature, MK::PartialMultisignature>,
+    pub recipient: NodeIndex,
+}
+
+/// Stores the current phase of a hash, and all relevant data useful in that phase:
+///
+/// * in the signature aggregation phase we store a signed hash and an aggregated partial
+/// multisignature under the hash
+/// * in the multisignature aggregation phase, we store the complete multisignature of the hash
+/// and the set of nodes about who we know that they have a multisignature
+/// * in the completed phase we only store the multisigned hash.
+enum HashState<'a, H: Signable, MK: MultiKeychain> {
+    SignatureAggregation {
+        signed: Signed<'a, Indexed<H>, MK>,
+        partial: PartiallyMultisigned<'a, H, MK>,
+    },
+    MultisignatureAggregation {
+        multisigned: Multisigned<'a, H, MK>,
+        confirmed: BoolNodeMap,
+    },
+    Completed {
+        _multisigned: Multisigned<'a, H, MK>,
+    },
+}
+
+impl<'a, H: Signable + Clone, MK: MultiKeychain> HashState<'a, H, MK> {
+    fn new(signed: Signed<'a, Indexed<H>, MK>, keychain: &'a MK) -> Self {
+        let partial = signed.clone()._into_partially_multisigned(keychain);
+        HashState::SignatureAggregation { signed, partial }
+    }
+
+    fn handle_signed_hash(
+        maybe_state: Option<Self>,
+        signed_indexed_hash: Signed<'a, Indexed<H>, MK>,
+        node_count: NodeCount,
+        keychain: &'a MK,
+        multisigned_hashes_tx: &UnboundedSender<Multisigned<'a, H, MK>>,
+    ) -> Option<Self> {
+        match maybe_state {
+            Some(HashState::SignatureAggregation {
+                signed,
+                mut partial,
+            }) => {
+                partial.add_signature(signed_indexed_hash);
+                match partial._try_into_complete(keychain) {
```

Comment:
This flow with returning sth on Err seems odd. Imho more natural would be to check `if keychain.is_complete(partial)` and then proceed with the Ok branch below.

Reply:
I would rather avoid calling `keychain.is_complete(partial)` and then unwrapping the result. This can be made explicit in type by making `PartiallyMultisigned<'a, T, MK>` and `enum` with two variants: `Incomplete{ signable: T, signature: MK::PartialMultisignature}` and `Complete(Multisigned<'a, T, MK>)`. Then instead of calling a conversion function, we can simply pattern match the enum. Then in fact the `HashState` can be simply replaced with `PartiallyMultisigned<'a, T, MK>`.